pkg/cmd/secret/set/set.go

```diff
@@ -0,0 +1,209 @@
+				}
+
+				if opts.Visibility != shared.All && opts.Visibility != shared.Private && opts.Visibility != shared.Selected {
+					return &cmdutil.FlagError{Err: errors.New(
+						"--visibility must be one of `all`, `private`, or `selected`")}
+				}
+
+				if opts.Visibility != shared.Selected && cmd.Flags().Changed("repos") {
+					return &cmdutil.FlagError{Err: errors.New(
+						"--repos only supported when --visibility='selected'")}
+				}
+
+				if opts.Visibility == shared.Selected && !cmd.Flags().Changed("repos") {
+					return &cmdutil.FlagError{Err: errors.New(
+						"--repos flag required when --visibility='selected'")}
+				}
+			} else {
+				if cmd.Flags().Changed("repos") {
+					opts.Visibility = shared.Selected
+				}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return setRun(opts)
+		},
+	}
+	cmd.Flags().StringVarP(&opts.OrgName, "org", "o", "", "List secrets for an organization")
+	cmd.Flags().StringVarP(&opts.Visibility, "visibility", "v", "private", "Set visibility for an organization secret: `all`, `private`, or `selected`")
+	cmd.Flags().StringSliceVarP(&opts.RepositoryNames, "repos", "r", []string{}, "List of repository names for `selected` visibility")
+	cmd.Flags().StringVarP(&opts.Body, "body", "b", "", "A value for the secret. Reads from STDIN if not specified.")
+
+	return cmd
+}
+
+func setRun(opts *SetOptions) error {
+	body, err := getBody(opts)
+	if err != nil {
+		return fmt.Errorf("did not understand secret body: %w", err)
+	}
+
+	c, err := opts.HttpClient()
+	if err != nil {
+		return fmt.Errorf("could not create http client: %w", err)
+	}
+	client := api.NewClientFromHTTP(c)
+
+	orgName := opts.OrgName
+
+	var baseRepo ghrepo.Interface
+	if orgName == "" {
+		baseRepo, err = opts.BaseRepo()
+		if err != nil {
+			return fmt.Errorf("could not determine base repo: %w", err)
+		}
+	}
+
+	var pk *PubKey
+	if orgName != "" {
+		pk, err = getOrgPublicKey(client, orgName)
+	} else {
+		pk, err = getRepoPubKey(client, baseRepo)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to fetch public key: %w", err)
+	}
+
+	eBody, err := box.SealAnonymous(nil, body, &pk.Raw, opts.RandomOverride)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt body: %w", err)
+	}
+
+	encoded := base64.StdEncoding.EncodeToString(eBody)
+
+	if orgName != "" {
+		err = putOrgSecret(client, pk, *opts, encoded)
+	} else {
+		err = putRepoSecret(client, pk, baseRepo, opts.SecretName, encoded)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to set secret: %w", err)
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+
+		if orgName == "" {
+			fmt.Fprintf(opts.IO.Out, "%s Set secret %s for %s\n", cs.SuccessIcon(), opts.SecretName, ghrepo.FullName(baseRepo))
+		} else {
+			fmt.Fprintf(opts.IO.Out, "%s Set secret %s for %s\n", cs.SuccessIcon(), opts.SecretName, orgName)
+		}
+	}
+
+	return nil
+}
+
+func validSecretName(name string) error {
```

Comment:
Nit: I think we could have gotten away without client-side validation. These rules may change in the future on the platform, and instead of syncing the logic here, we could have allowed any value and let the server decide whether to accept them.